Update Ingress Docs about new ingress mode (#840)
linkerd/website: linkerd.io/content/2/tasks/using-ingress.md

description = "Linkerd works alongside your ingress controller of choice."
+++

When your ingress controller is injected, The load balancing decision would

> be taken by the Ingress Controller and the proxy just routes it to the

Reviewer: perhaps clarify "Linkerd proxy" instead of just "proxy"

choosen target Pod IP by the Ingress controller. This means that Linkerd functionality
like Service Profiles, Traffic Splits would not work as the current discovery is

> based on the Service IP. TCP mTLS and Load Balancing should still work however.

Reviewer: s/should/will

This is useful in cases where you want Linkerd to honour the load balancing
decision taken by the Ingress Controller for things like Sticky Sessions, etc.

If you're planning on injecting Linkerd into your ingress controller's pods
and want Linkerd functionality like Service Profiles, Traffic Splits, etc,
there is additional configuration required to make the proxy run in `ingress` mode,
where Linkerd discovers services based on the `:authority` or `Host` header.
This allows Linkerd to understand what service a request is destined for
without being dependent on DNS or IPs.

The Ingress controller deployment can be made to run in `ingress` mode by adding
the following annotation.

```bash
kubectl get deployment <ingress-controller> -n <ingress-namespace> -o yaml | linkerd inject --ingress - | kubectl apply -f -
```

This can be verified by checking if the `linkerd-proxy` has the relevant environment
variable set by the `proxy-injector`.

```bash

> kubectl describe pod/<ingress-pod> | grep LINKERD2_PROXY_INGRESS_MODE

Reviewer: can we look at the annotation in the pod spec of the deployment instead? it would be nice to avoid referencing proxy env variables in public facing documentation

description = "Linkerd works alongside your ingress controller of choice."
+++

> As of Linkerd version 2.9, There are two ways in which Linkerd proxy

Reviewer: "There" should not be capitalized

> As of Linkerd version 2.9, There are two ways in which Linkerd proxy

Reviewer: "...in which *the* Linkerd proxy..."

description = "Linkerd works alongside your ingress controller of choice."
+++

From Linkerd `v2.9`, There are two ways in which Linkerd proxy
can be ran with your Ingress Controller.

## Default Mode

When the ingress controller is injected, it makes the balancing

> decision by choosing the target Pod IP and the inkerd proxy just routes it.

Reviewer: "...is injected with the `linkerd.io/inject: enabled` annotation, the Linkerd proxy will honor load balancing decisions made by the ingress controller instead of applying its own EWMA load balancing.  This also means that the Linkerd proxy will not use Service Profiles for this traffic and therefore will not expose per-route metrics or do traffic splitting."

> When the ingress controller is injected, The load balancing decision would
> be taken by the Ingress Controller and the Linkerd proxy just routes it to the
> choosen target Pod IP by the Ingress controller. This means that Linkerd functionality

Reviewer 2: This first sentence felt a little wordy, but not a blocker:

```suggestion
When the ingress controller is injected, it makes the load balancing decision by choosing the target pod IP and the Linkerd proxy routes it. This means that Linkerd functionality
```

> like Service Profiles, Traffic Splits would not work as the current discovery is

Reviewer 2: ```suggestion
like Service Profiles, Traffic Splits would not work as its discovery is
```

> based on the Service IP. TCP mTLS and Load Balancing will still work however.

Reviewer 2: It may be worth noting here that the load balancing that is being honored is that of the ingress controller and not the proxy's load balancing.

Author: I'm considering just removing `load balancing` here as it's already understood from the previous sentences?

Reviewer 2: Yep that sounds good!

This is useful in cases where you want Linkerd to honour the load balancing
decision taken by the Ingress Controller for things like Sticky Sessions, etc.

If your Ingress controller is injected with no other configuration specific to

> ingress, The Linkerd proxy runs in the default mode.

Reviewer 2: ```suggestion
ingress, the Linkerd proxy runs in the default mode.
```

## Proxy Ingress Mode

If you want Linkerd functionality like Service Profiles, Traffic Splits, etc,
there is additional configuration required to make the Ingress controller's Linkerd

> proxy run in `ingress` mode. This allows Linkerd to discover services based on the

Reviewer: ".. `ingress` mode.  This causes Linkerd to route requests based on their `:authority`, `Host`, or `l5d-dst-override` header instead of their original destination which allows Linkerd to perform its own load balancing and use Service Profiles to expose per-route metrics and enable traffic splitting."

> If you want Linkerd functionality like Service Profiles, Traffic Splits, etc,
> there is additional configuration required to make the Ingress controller's Linkerd
> proxy run in `ingress` mode, where Linkerd discovers services based on the
> `:authority` or `Host` header. This allows Linkerd to understand what service a
> request is destined for without being dependent on DNS or IPs.

Reviewer 2: ```suggestion
If you want Linkerd functionality like Service Profiles, Traffic Splits, etc, there is additional configuration required to make the Ingress controller’s Linkerd proxy run in ingress mode. This allows Linkerd to discover services based on `:authority` or `Host` headers and understand what service a request is destined for without being dependent on DNS or IPs
```
